File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/scheduler/SchedulerTest.kt
package io.github.jkjamies.automationagent.scheduler

import io.github.jkjamies.automationagent.ingest.Envelope
import io.github.jkjamies.automationagent.ingest.Kind
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeout
import java.time.Instant

class SchedulerTest : BehaviorSpec({
    Given("valid and invalid cron specs") {
        When("adding them") {
            Then("valid specs register and an invalid spec throws") {
                val s = Scheduler({ })
                s.add("0 9 * * *", Kind.CRON_DAILY)
                s.add("0 9 * * 1", Kind.CRON_WEEKLY)
                s.entries() shouldBe 2
                shouldThrow<IllegalArgumentException> { s.add("not a cron spec", Kind.CRON_DAILY) }
            }
        }
    }

    Given("an @every schedule") {
        When("started") {
            Then("it fires and can be stopped") {
                val fired = CompletableDeferred<Kind>()
                val s = Scheduler({ e -> fired.complete(e.kind) })
                s.add("@every 1s", Kind.CRON_DAILY)
                s.start()
                try {
                    val kind = withTimeout(3000) { fired.await() }
                    kind shouldBe Kind.CRON_DAILY
                } finally {
                    s.stop()
                }
            }
        }
    }

    Given("a fixed clock") {
        When("triggering directly") {
            Then("it emits a scheduler envelope with the clock's timestamp") {
                var captured: Envelope? = null
                val fixed = Instant.ofEpochSecond(1718870400)
                val s = Scheduler({ e -> captured = e }, now = { fixed })
                s.trigger(Kind.CRON_WEEKLY)
                val got = captured.shouldNotBeNull()
                got.kind shouldBe Kind.CRON_WEEKLY
                got.source shouldBe "scheduler"
                got.receivedAt shouldBe fixed
            }
        }
    }
})

File: go/internal/scheduler/scheduler_test.go
package scheduler

import (
	"testing"
	"time"

	"github.com/jkjamies/automation-agent/internal/ingest"
)

func TestAddValidAndInvalid(t *testing.T) {
	s := New(func(ingest.Envelope) {})
	if err := s.Add("0 9 * * *", ingest.KindCronDaily); err != nil {
		t.Errorf("valid daily spec: %v", err)
	}
	if err := s.Add("0 9 * * 1", ingest.KindCronWeekly); err != nil {
		t.Errorf("valid weekly spec: %v", err)
	}
	if s.Entries() != 2 {
		t.Errorf("entries = %d, want 2", s.Entries())
	}
	if err := s.Add("not a cron spec", ingest.KindCronDaily); err == nil {
		t.Error("expected error for invalid spec")
	}
}

func TestStartFiresAndStop(t *testing.T) {
	ch := make(chan ingest.Envelope, 1)
	s := New(func(e ingest.Envelope) {
		select {
		case ch <- e:
		default:
		}
	})
	if err := s.Add("@every 1s", ingest.KindCronDaily); err != nil {
		t.Fatalf("Add: %v", err)
	}
	s.Start()
	defer s.Stop()

	select {
	case e := <-ch:
		if e.Kind != ingest.KindCronDaily {
			t.Errorf("fired kind = %q", e.Kind)
		}
	case <-time.After(3 * time.Second):
		t.Fatal("schedule did not fire within 3s")
	}
}

func TestTriggerEmitsEnvelope(t *testing.T) {
	var got ingest.Envelope
	s := New(func(e ingest.Envelope) { got = e })
	fixed := time.Unix(1718870400, 0)
	s.now = func() time.Time { return fixed }

	s.trigger(ingest.KindCronWeekly)

	if got.Kind != ingest.KindCronWeekly {
		t.Errorf("kind = %q, want %q", got.Kind, ingest.KindCronWeekly)
	}
	if got.Source != "scheduler" {
		t.Errorf("source = %q", got.Source)
	}
	if !got.ReceivedAt.Equal(fixed) {
		t.Errorf("ReceivedAt = %v, want %v", got.ReceivedAt, fixed)
	}
}
